[Scalars] Remove Scalar from return types of functions.

gchanan opened the pull request:
This also changes the scalar-from-tensor conversion from explicit to implicit.
This allows code such as t.fill_(t.sum()) to continue working, while still
documenting that only a scalar is supported.

gchanan commented:
Maybe we should add direction scalar conversion functions to tensor; e.g. instead of calling Scalar(t).toDouble(), it's a bit more convenient to do t.toScalarDouble().

gchanan commented:
This addresses https://github.com/zdevito/ATen/issues/116.

zdevito commented on src/ATen/Scalar.h:
```diff
@@ -21,7 +21,7 @@ class AT_API Scalar {
   Scalar() : Scalar(int64_t(0)) {}
 
-  explicit Scalar(const detail::TensorBase & t)
+  Scalar(const detail::TensorBase & t)
```
The explicit prevents horrible overload problems when Scalar and Tensor can appear in the same position in a function list. I believe it only spawns when there is _another_ argument that also needs some implicit conversion.  I am hesitant to remove explicit until we remove all the cases where Scalar and Tensor can appear in the same argument list.

gchanan replied:
I thought about this but figured because nothing got tripped up we were okay; I'll look into this a bit more.

killeent replied:
Yes I would second not getting rid of explicit constructors. Implicit constructors are really quite annoying.

gchanan replied:
Ok, how about this: we get rid of the implicit conversion and add a Tensor overload for functions like fill, where we want x.fill(y.mean()) to continue working (note: it doesn't work because those functions only have a Scalar argument now).  I don't know if we figure out that set of functions that programmatically or we mark them in the cwrap somehow.  That is against the spirit of https://github.com/zdevito/ATen/issues/117, but we can fix it when we address that issue.

gchanan replied:
I looked through the functions and here are functions where there exists a Scalar argument that doesn't have a Tensor equivalent overload already, grouped by related scalar argument types:

**fills:**
- fill_
- maskedFill_
- indexFill_

**range/min/max:**
- range
- arange
- linspace
- logspace
- histc
- clamp
- clamp_

**exponent/p-norm:**
- norm
- renorm
- renorm_
- dist

**scaling factor/weight:**
- lerp
- lerp_
- add
- add_
- sub
- sub_
- addmm
- addmm_
- addmv
- addmv_
- addr
- addr_
- addbmm
- addbmm_
- baddbmm
- baddbmm_
- baddbmm_
- addcmul
- addcmul_
- addcdiv
- addcdiv_

**tolerance:**
- pstrf

gchanan replied:
I think the most straightforward way of implementing this is just to rely on the handle_zero_dim code; we'll add an option 'only_zero_dim' or similar which means the function should just throw an exception if something that isn't zero_dim is passed in.  Then it's just a matter of choosing which overloads we want to define in Declarations.cwrap.

I'd argue for just the fill functions at this point: I don't really see a use case for the exponent/p-norm, scaling factor/weight, or tolerance being the result of some tensor calculation (at least I haven't used them as such).  I have used them in ranges, but ranges/min/max take multiple values and handle_zero_dim doesn't handle that yet anyway.

zdevito commented on src/ATen/function_wrapper.py:
```diff
@@ -918,6 +918,8 @@ def emit_body(env, option):
                 body.append(CodeTemplate(return_tensor).substitute(
                     env, arg_name=call, maybe_scalar=maybe_scalar))
+            elif ret['type'] == 'accreal' or ret['type'] == 'real':
+                body.append('return scalarTensor({});'.format(call))
```
For 'accreal', will this be the right type? For instance, for Byte this will make a ByteTensor, but shouldn't it be a bigger type for accreal?

gchanan replied:
you are correct, I misread the definition of scalarTensor.

gchanan replied:
actually I think we should always return a tensor of the real type for both real and accreal specified returns.

Otherwise consider the following two calls and their respective return types on CPU/GPU for a FloatTensor x (assuming we define real/accreal as it is defined in TH/THC):

| Device | x.sum() |  x.sum(0) |
| ------- | ------- | ---------- |
| CPU     | Double | Float |
| GPU     | Float     | Float |

This is pretty confusing and also inconsistent with what we currently do with the PyTorch Variable API (they always return FloatTensors).

For people worried about precision we should accept a dtype arugment, which we are planning to add to PyTorch as well.

zdevito commented on src/ATen/test/basic.cpp:
```diff
@@ -33,11 +33,11 @@ static void test(Type & type) {
     ASSERT(12 == b.numel());
     std::cout << b.dot(b) << std::endl;
-    ASSERT(b.dot(b).toDouble() == 12);
+    ASSERT(Scalar(b.dot(b)).toDouble() == 12);
```
For direct users of ATen this is pretty annoying. Can we add the helpers toDouble/toFloat, etc. directly to Tensor (in this case they should just do Scalar(fo).toX())?

gchanan replied:
I agree this is a good idea, see my message above.  I'm a bit concerned that users would think toFloat would be defined as toType(kFloat), not Scalar(*this).toFloat().  What do you think?  We could alternatively call this something like toFloatScalar() or (if people think that would return a Scalar) toCFloat()?

zdevito replied:
Yeah, we should be specific, maybe `toNativeFloat()`? `toFloatScalar()` does sound like it will return a `Scalar`.

gchanan replied:
I have a slight preference for toCFloat() over toNativeFloat() because it's shorter and native already has a meaning in ATen (native functions), but I'm happy to go with it if that's your preference.

zdevito replied:
toCFloat is good

gchanan commented:
I may need this to move matmul to C++ because the 1-d X 1-d case currently returns a scalar.

gchanan commented:
I made the following changes which I think addresses the review comments:

- Scalar no longer has an implicit conversion.  Instead, for fill functions we explicitly define Tensor overloads (that throw if non-0-dim tensors are passed in)
- We introduce toCFloat(), toCDouble() etc. for simpler C-scalar conversions and change the tests to use these.

zdevito commented on src/ATen/function_wrapper.py:
```diff
@@ -761,6 +775,8 @@ def emit_body(env, option):
         body = []
         body += handle_sparse(env, option)
         body += handle_zero_dim(env, option)
+        if only_zero_dim(env, option):
```
Why does this cause a return?

gchanan replied:
the generated code is:
```
Tensor & CPUFloatType::m_fill_(Tensor & self, const Tensor & value) const {
    if(value.dim() == 0) {
        return static_cast<const Type*>(this)->m_fill_(self, Scalar(value));
    }
    else {
        runtime_error("fill_ only supports a 0-dimensional value tensor, but got tensor "
                      "with %" PRId64 " dimension(s)", value.dim());
    }
}
```

so anything else is unreachable.  Also, the further generated code if we didn't return would be invalid as of now; we could change it to not be, but since it's unreachable it didn't seem worth the effort.

zdevito replied:
ok I just have some small nits about this to make it more understandable:
I think we should avoid templates that are not complete statements/expressions, so instead of `else ...`, can we just repeat the check `if(what.dim() != 0) { ... }`. This will be more robust to future modifications. 

Can we also make all the only_zero_dim handling happen in one function? `only_zero_dim` -> `handle_only_zero_dim`, and merge all the only emission code  into it:

```
zero_dim_check = handle_only_zero_dim(env, option)
if zero_dim_check is not None:
   # code below zero dim check is an asset so we do not need to generate the rest.
   body += zero_dim_check
   return
```

zdevito commented:
I think this approach is good  since it avoids tricky overload options for now. Eventually we will add some way to represent lightweight allocation of a 0-dim tensor and can remove all the remaining Scalar cases. I posted one comment about how the zero-dim check works.
